sqlmesh/core/plan/definition.py

```diff
@@ -308,16 +287,34 @@ def loaded_snapshot_intervals(self) -> t.List[LoadedSnapshotIntervals]:
                 continue
             loaded_snapshots.append(LoadedSnapshotIntervals.from_snapshot(snapshot))
             for downstream_indirect in self.indirectly_modified.get(snapshot.name, set()):
-                downstream_snapshot = self.context_diff.snapshots[downstream_indirect]
+                downstream_snapshot = self._snapshot_mapping[downstream_indirect]
                 # We don't want to display indirect non-breaking since to users these are effectively no-op changes
                 if downstream_snapshot.is_indirect_non_breaking:
                     continue
                 loaded_snapshots.append(LoadedSnapshotIntervals.from_snapshot(downstream_snapshot))
         return loaded_snapshots
 
+    @property
+    def has_changes(self) -> bool:
+        modified_names = (
+            set().union(
+                *[
+                    self.context_diff.added,
+                    self.context_diff.removed,
+                    self.context_diff.modified_snapshots,
+                ]
+            )
+            - self.ignored_snapshot_names
+        )
+        return (
+            self.context_diff.is_new_environment
+            or self.context_diff.is_unfinalized_environment
+            or bool(modified_names)
+        )
+
     def is_new_snapshot(self, snapshot: Snapshot) -> bool:
         """Returns True if the given snapshot is a new snapshot in this plan."""
-        return snapshot.snapshot_id in self.context_diff.new_snapshots
+        return snapshot.snapshot_id in [s.snapshot_id for s in self.new_snapshots]
```

Comment:
I think we call `is_new_snapshot` quite a bit, including inside other loops iterating over snapshots. This may potentially be `O(N^2)`. Perhaps worth caching and using set